File: Javascript/DCP02_array_product.js
let WITH_DIVISION = 1;
let NO_DIVISION = 0;

function array_product(arr, option)
{
    let output_array = [];

    // Linear time solution.
    if(option === WITH_DIVISION)
    {
        let prod = 1;

        for(let i = 0; i < arr.length; i++)
        {
            prod = prod * arr[i];
        }

        for(let i = 0; i < arr.length; i++)
        {
            output_array[i] = prod / input_array[i];
        }
    }

    // Quadratic time solution.
    else
    {
        for(let i = 0; i < arr.length; i++)
        {
            let prod = 1;
            for(let j = 0; j < arr.length; j++)
            {
                if(j != i)
                {
                    prod = prod * arr[j];
                }
            }

            output_array[i] = prod;
        }
    }

    return output_array;
}

let input_array = [1, 2, 3, 4, 5];

// [120, 60, 40, 30, 24]
console.log(array_product(input_array, WITH_DIVISION)); 
console.log(array_product(input_array, NO_DIVISION)); 

input_array = [3, 2, 1];

// [2, 3, 6]
console.log(array_product(input_array, WITH_DIVISION)); 
console.log(array_product(input_array, NO_DIVISION)); 

File: Python/DCP02_array_product.py
WITH_DIVISION = 0
NO_DIVISION = 1

def array_product(arr, option):
    output = [None] * len(arr)
    if(option is WITH_DIVISION):
        prod = 1
        for elem in arr:
            prod = prod * elem
        for i in range(0, len(arr)):
            output[i] = prod / arr[i]
    else:
        for i in range(0, len(arr)):
            prod = 1
            for j in range(0, len(arr)):
                if(i != j):
                    prod = prod * arr[j]
            output[i] = prod
    return output

arr = [1, 2, 3, 4, 5]
print(array_product(arr, WITH_DIVISION)) # [120, 60, 40, 30, 24]
print(array_product(arr, NO_DIVISION)) # [120, 60, 40, 30, 24]

arr = [3, 2, 1]
print(array_product(arr, WITH_DIVISION)) # [2, 3, 6]
print(array_product(arr, NO_DIVISION)) # [2, 3, 6]
